### ISSUE TITLE
[v4] Dynamic height with FlatList not scrolling.

### ISSUE THREAD

**Issue by alexisloiselle** [gorhom/react-native-scroll-bottom-sheet#658]
# Bug

When using the `useBottomSheetDynamicSnapPoints` together with `BottomSheetFlatList`, it's not possible to scroll the list anymore.

https://user-images.githubusercontent.com/16881226/134593038-28bd8fc6-841c-4e91-a380-b0a231b7e70d.MP4

## Environment info

| Library                         | Version |
| ------------------------------- | ------- |
| @gorhom/bottom-sheet            | 4.0.3   |
| react-native                    | 0.64.2   |
| react-native-reanimated         | 2.2.2   |
| react-native-gesture-handler    | 1.10.3   |

## Steps To Reproduce

1. add a `BottomSheet` with the `animatedSnapPoints`, `animatedContentHeight` and `animatedHandleHeight` provided by `useBottomSheetDynamicSnapPoints`
2. set the `handleContentLayout` to the child
3. add a big enough list to overflow the screen

Describe what you expected to happen:

1. it's not possible to scroll anymore

## Reproducible sample code

Here's a Snack to reproduce:
https://snack.expo.dev/@alexisloiselle/dynamic-height-with-flatlist-not-scrolling

**Comment by VyacheslavAA** [comment]
Hello @alexisloiselle !

I had the same issue with `BottomSheetScrollView`.

As a temporary solution you can use `BottomSheetScrollView` and change `height` property for `BottomSheetScrollView` using `useAnimatedStyle` to constraint height of `BottomSheetScrollView` .
Also, to prevent infinity loop for `handleContentLayout` you need to add additional `View` with onLayout handler as `BottomSheetScrollView` child.

Here is example:

```
import BottomSheet, {
  BottomSheetScrollView,
  useBottomSheetDynamicSnapPoints
} from '@gorhom/bottom-sheet';
import React, { useMemo } from 'react';
import { useWindowDimensions, View } from 'react-native';
import { useAnimatedStyle } from 'react-native-reanimated';

export const YourComponent = () => {
  const { height: windowHeight } = useWindowDimensions();

  const initialSnapPoints = useMemo(() => [350, 'CONTENT_HEIGHT'], []);
  const {
    animatedHandleHeight,
    animatedSnapPoints,
    animatedContentHeight,
    handleContentLayout,
  } = useBottomSheetDynamicSnapPoints(initialSnapPoints);

  const scrollViewAnimatedStyles = useAnimatedStyle(() => {
    const contentHeight = animatedContentHeight.value;
    const handleHeight = animatedHandleHeight.value;
    const bottomSheetHeight = handleHeight + contentHeight;

    return {
      height:
        bottomSheetHeight > windowHeight
          ? windowHeight - handleHeight
          : bottomSheetHeight,
    };
  });

  return (
    <BottomSheet
      snapPoints={animatedSnapPoints}
      handleHeight={animatedHandleHeight}
      contentHeight={animatedContentHeight}
    >
      <BottomSheetScrollView style={scrollViewAnimatedStyles}>
        <View onLayout={handleContentLayout}>{/* Children here */}</View>
      </BottomSheetScrollView>
    </BottomSheet>
  );
};
```

**Comment by tslater** [comment]
@VyacheslavAA Thanks for sharing your workaround. Just to make sure I understand correctly, you're using the window height? Does that mean this only works in the limited case that your bottom sheet goes full-screen?

**Comment by pronevich** [comment]
Same issue on Android
@VyacheslavAA thanks, your solution works.

**Comment by gorhom** [comment]
dynamic snap points currently working only with views, there are still some work to get it working with scrollables

**Comment by pronevich** [comment]
This issue still annoying) I have looked at the source code, but can't figure out how to fix it.

**Comment by tslater** [comment]
@pronevich I agree. @gorhom, anyway to re-open this? Do you have any ideas here?

**Comment by hisothreed** [comment]
any updates here? Has anyone here managed to fix it or find a workaround for this issue?

**Comment by hisothreed** [comment]
actually found a workaround:

```
  const initialSnapPoints = useMemo(() => ['CONTENT_HEIGHT'], []);
  const {
    animatedHandleHeight,
    animatedSnapPoints,
    animatedContentHeight,
    handleContentLayout,
  } = useBottomSheetDynamicSnapPoints(initialSnapPoints);

  const insets = useSafeAreaInsets();
  const maxHeight = heightPercentageToDP(100) - +animatedHandleHeight.value;
  const contentContainerStyle = {
    paddingBottom: insets.bottom + animatedHandleHeight.value,
  };
  const scrollStyle = {
    maxHeight,
  };
  const onLayout = e => {
    const {height} = e.nativeEvent.layout;
    if (height > maxHeight) {
      return handleContentLayout({
        nativeEvent: {
          layout: {
            height: maxHeight,
          },
        },
      });
    }
    return handleContentLayout(e);
  };

...
<BottomSheetFlatList<IBankAccount>
          onLayout={onLayout}
          style={contentContainerStyle}
          contentContainerStyle={scrollStyle}
          data={accounts}
          renderItem={renderItem}
          keyExtractor={keyExtractor}
/>
...
```

TL;DR:  take whatever content height the list is throwing ```onLayout``` event trigger and give it to the bottom sheet hook unless the height of the list is larger than the screen height, if so then limit the height of the bottom sheet list to the height of the screen minus the height of the stuff around the flat list.

**Comment by doougui** [comment]
> Hello @alexisloiselle !
> 
> I had the same issue with `BottomSheetScrollView`.
> 
> As a temporary solution you can use `BottomSheetScrollView` and change `height` property for `BottomSheetScrollView` using `useAnimatedStyle` to constraint height of `BottomSheetScrollView` . Also, to prevent infinity loop for `handleContentLayout` you need to add additional `View` with onLayout handler as `BottomSheetScrollView` child.
> 
> Here is example:
> 
> ```
> import BottomSheet, {
>   BottomSheetScrollView,
>   useBottomSheetDynamicSnapPoints
> } from '@gorhom/bottom-sheet';
> import React, { useMemo } from 'react';
> import { useWindowDimensions, View } from 'react-native';
> import { useAnimatedStyle } from 'react-native-reanimated';
> 
> export const YourComponent = () => {
>   const { height: windowHeight } = useWindowDimensions();
> 
>   const initialSnapPoints = useMemo(() => [350, 'CONTENT_HEIGHT'], []);
>   const {
>     animatedHandleHeight,
>     animatedSnapPoints,
>     animatedContentHeight,
>     handleContentLayout,
>   } = useBottomSheetDynamicSnapPoints(initialSnapPoints);
> 
>   const scrollViewAnimatedStyles = useAnimatedStyle(() => {
>     const contentHeight = animatedContentHeight.value;
>     const handleHeight = animatedHandleHeight.value;
>     const bottomSheetHeight = handleHeight + contentHeight;
> 
>     return {
>       height:
>         bottomSheetHeight > windowHeight
>           ? windowHeight - handleHeight
>           : bottomSheetHeight,
>     };
>   });
> 
>   return (
>     <BottomSheet
>       snapPoints={animatedSnapPoints}
>       handleHeight={animatedHandleHeight}
>       contentHeight={animatedContentHeight}
>     >
>       <BottomSheetScrollView style={scrollViewAnimatedStyles}>
>         <View onLayout={handleContentLayout}>{/* Children here */}</View>
>       </BottomSheetScrollView>
>     </BottomSheet>
>   );
> };
> ```

This only works the first time that the modal is open. If we close and re-open the modal it doesn't scroll anymore

**Comment by pronevich** [comment]
Issue still here

**Comment by vitorsilvalima** [comment]
yes, this is still happening =/

**Comment by mattgabor** [comment]
Still an issue

**Comment by FadiAboMsalam** [comment]
why was this closed ?!!!!

**Comment by aymenbuu** [comment]
please fix

EDIT:

Just found there is a BottomSheetFlatList component, as well as ScrollView/ VirtualizedList etc,, incase some people like me didn't know. This fixed my issues.

**Comment by JohnBerd** [comment]
@gorhom please re-open this issue please, it is still not fixed

**Comment by izakfilmalter** [comment]
Would really like a solution to this. This would be so clutch for me.

**Comment by janroures** [comment]
Please re-open this. ScrollView is not scrolling full content height, might be related to this

**Comment by fukemy** [comment]
no solution now

**Comment by alexdobson** [comment]
Would be great to have a solution for this!

**Comment by jc578271** [comment]
fixed by creating your own dynamic snap points hook:

```typescript
const animatedContentHeight = useSharedValue(0);
const animatedSnapPoints = useDerivedValue(() => {
  return ['25%', animatedContentHeight.value + 24];  // add handleHeight
});

// ...

 <BottomSheetModal
    ref={bottomSheetRef}
    snapPoints={animatedSnapPoints}
    // dont set contentHeight prop
>

  <BottomSheetFlatList
    onLayout={e => {
      e.currentTarget.getInnerViewRef().measure((x, y, w, h) => { // measure inner view of FlatList
        animatedContentHeight.value = h;
      });
    }}
    data={['1', '2', '3', '4', '5', '6']}
    renderItem={({item}) => (
      <View>
        <Text>{item}</Text>
      </View>
    )}
  />
</BottomSheetModal>
```

**Comment by anggadarkprince** [comment]
> Hello @alexisloiselle !
> 
> I had the same issue with `BottomSheetScrollView`.
> 
> As a temporary solution you can use `BottomSheetScrollView` and change `height` property for `BottomSheetScrollView` using `useAnimatedStyle` to constraint height of `BottomSheetScrollView` . Also, to prevent infinity loop for `handleContentLayout` you need to add additional `View` with onLayout handler as `BottomSheetScrollView` child.
> 
> Here is example:
> 
> ```
> import BottomSheet, {
>   BottomSheetScrollView,
>   useBottomSheetDynamicSnapPoints
> } from '@gorhom/bottom-sheet';
> import React, { useMemo } from 'react';
> import { useWindowDimensions, View } from 'react-native';
> import { useAnimatedStyle } from 'react-native-reanimated';
> 
> export const YourComponent = () => {
>   const { height: windowHeight } = useWindowDimensions();
> 
>   const initialSnapPoints = useMemo(() => [350, 'CONTENT_HEIGHT'], []);
>   const {
>     animatedHandleHeight,
>     animatedSnapPoints,
>     animatedContentHeight,
>     handleContentLayout,
>   } = useBottomSheetDynamicSnapPoints(initialSnapPoints);
> 
>   const scrollViewAnimatedStyles = useAnimatedStyle(() => {
>     const contentHeight = animatedContentHeight.value;
>     const handleHeight = animatedHandleHeight.value;
>     const bottomSheetHeight = handleHeight + contentHeight;
> 
>     return {
>       height:
>         bottomSheetHeight > windowHeight
>           ? windowHeight - handleHeight
>           : bottomSheetHeight,
>     };
>   });
> 
>   return (
>     <BottomSheet
>       snapPoints={animatedSnapPoints}
>       handleHeight={animatedHandleHeight}
>       contentHeight={animatedContentHeight}
>     >
>       <BottomSheetScrollView style={scrollViewAnimatedStyles}>
>         <View onLayout={handleContentLayout}>{/* Children here */}</View>
>       </BottomSheetScrollView>
>     </BottomSheet>
>   );
> };
> ```

Some issue occur when add inputs, when keyboard shows up, it cannot scrolled down, based on solution above we can manage the height by subtract current height with keyboard height to maintain the scroll reach the bottom.

```js
export const useKeyboard = () => {
  const [keyboardHeight, setKeyboardHeight] = useState(0);

  useEffect(() => {
    function onKeyboardDidShow(e: KeyboardEvent) {
      setKeyboardHeight(e.endCoordinates.height);
    }

    function onKeyboardDidHide() {
      setKeyboardHeight(0);
    }

    const showSubscription = Keyboard.addListener(
      'keyboardDidShow',
      onKeyboardDidShow,
    );
    const hideSubscription = Keyboard.addListener(
      'keyboardDidHide',
      onKeyboardDidHide,
    );
    return () => {
      showSubscription.remove();
      hideSubscription.remove();
    };
  }, []);

  return keyboardHeight;
};

```

use keyboard hooks to calculate the height
```js
const keyboardHeight = useKeyboard();
const bottomOffside = 120;

const scrollViewAnimatedStyles = useAnimatedStyle(() => {
    const contentHeight = animatedContentHeight.value;
    const handleHeight = animatedHandleHeight.value + bottomOffside; // add offside if necessary
    const bottomSheetHeight = handleHeight + contentHeight;
    return {
      height:
        bottomSheetHeight > windowHeight
          ? windowHeight - handleHeight - keyboardHeight // subtract with keyboard height
          : bottomSheetHeight - keyboardHeight, // subtract with keyboard height
    };
  });

```

**Comment by manan-schmooze** [comment]
Still facing the issue

**Comment by VictorPulzz** [comment]
@gorhom Why you don't support your lib?

**Comment by ororsatti** [comment]
https://github.com/gorhom/react-native-bottom-sheet/discussions/1363 a scrollable dynamically size drawer

**Comment by MensurRasic** [comment]
Still having this issue...
